// @pxp-nd/core/dist/sockets/basic-sockets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.message = exports.disconnect = exports.connectClient = exports.usersConnected = void 0;
const typeorm_1 = require("typeorm");
const UserIo_1 = __importDefault(require("./entities/UserIo"));
exports.usersConnected = [];
const connectClient = (client, io) => {
    exports.usersConnected.push(client.id);
    typeorm_1.getManager().save(UserIo_1.default, {
        clientId: client.id,
        userId: 1,
    }).then(data => console.log(data)).catch(console.error);
};
exports.connectClient = connectClient;
const disconnect = (client, io) => {
    client.on('disconnect', () => {
        console.log('Cliente desconectado');
        const clientIndex = exports.usersConnected.indexOf(client.id);
        if (clientIndex > -1) {
            exports.usersConnected.splice(clientIndex, 1);
        }
        typeorm_1.getManager().delete(UserIo_1.default, {
            clientId: client.id,
        }).then(data => console.log(data)).catch(console.error);
        io.emit('active-users', exports.usersConnected.toString());
    });
};
exports.disconnect = disconnect;
const message = (client, io) => {
    client.on('message', (payload) => {
        console.log('Message recived', payload.from, payload.message);
        io.emit('new-message', payload);
    });
};
exports.message = message;
