// @pxp-nd/common/dist/entities/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.entities = exports.Location = exports.Currency = exports.WordKey = exports.User = exports.UiTransaction = exports.Ui = exports.Type = exports.TranslationGroup = exports.Translate = exports.Transaction = exports.Template = exports.Subtype = exports.Subsystem = exports.Session = exports.ScriptVersion = exports.Role = exports.ReportGroup = exports.Report = exports.Person = exports.Notification = exports.Message = exports.Log = exports.Language = exports.HistoryFile = exports.Group = exports.GlobalData = exports.FileType = exports.File = exports.ConfigValue = exports.Config = exports.ChatUser = exports.ChatType = exports.Chat = exports.CatalogValue = exports.Catalog = exports.Branch = exports.AccountStatusType = exports.AccountStatus = void 0;
const AccountStatus_1 = __importDefault(require("./AccountStatus"));
exports.AccountStatus = AccountStatus_1.default;
const AccountStatusType_1 = __importDefault(require("./AccountStatusType"));
exports.AccountStatusType = AccountStatusType_1.default;
const Branch_1 = __importDefault(require("./Branch"));
exports.Branch = Branch_1.default;
const Catalog_1 = __importDefault(require("./Catalog"));
exports.Catalog = Catalog_1.default;
const CatalogValue_1 = __importDefault(require("./CatalogValue"));
exports.CatalogValue = CatalogValue_1.default;
const Chat_1 = __importDefault(require("./Chat"));
exports.Chat = Chat_1.default;
const ChatType_1 = __importDefault(require("./ChatType"));
exports.ChatType = ChatType_1.default;
const ChatUser_1 = __importDefault(require("./ChatUser"));
exports.ChatUser = ChatUser_1.default;
const Config_1 = __importDefault(require("./Config"));
exports.Config = Config_1.default;
const ConfigValue_1 = __importDefault(require("./ConfigValue"));
exports.ConfigValue = ConfigValue_1.default;
const File_1 = __importDefault(require("./File"));
exports.File = File_1.default;
const FileType_1 = __importDefault(require("./FileType"));
exports.FileType = FileType_1.default;
const GlobalData_1 = __importDefault(require("./GlobalData"));
exports.GlobalData = GlobalData_1.default;
const Group_1 = __importDefault(require("./Group"));
exports.Group = Group_1.default;
const HistoryFile_1 = __importDefault(require("./HistoryFile"));
exports.HistoryFile = HistoryFile_1.default;
const Language_1 = __importDefault(require("./Language"));
exports.Language = Language_1.default;
const Log_1 = __importDefault(require("./Log"));
exports.Log = Log_1.default;
const Message_1 = __importDefault(require("./Message"));
exports.Message = Message_1.default;
const Notification_1 = __importDefault(require("./Notification"));
exports.Notification = Notification_1.default;
const Person_1 = __importDefault(require("./Person"));
exports.Person = Person_1.default;
const Report_1 = __importDefault(require("./Report"));
exports.Report = Report_1.default;
const ReportGroup_1 = __importDefault(require("./ReportGroup"));
exports.ReportGroup = ReportGroup_1.default;
const Role_1 = __importDefault(require("./Role"));
exports.Role = Role_1.default;
const ScriptVersion_1 = __importDefault(require("./ScriptVersion"));
exports.ScriptVersion = ScriptVersion_1.default;
const Session_1 = __importDefault(require("./Session"));
exports.Session = Session_1.default;
const Subsystem_1 = __importDefault(require("./Subsystem"));
exports.Subsystem = Subsystem_1.default;
const Subtype_1 = __importDefault(require("./Subtype"));
exports.Subtype = Subtype_1.default;
const Template_1 = __importDefault(require("./Template"));
exports.Template = Template_1.default;
const Transaction_1 = __importDefault(require("./Transaction"));
exports.Transaction = Transaction_1.default;
const Translate_1 = __importDefault(require("./Translate"));
exports.Translate = Translate_1.default;
const TranslationGroup_1 = __importDefault(require("./TranslationGroup"));
exports.TranslationGroup = TranslationGroup_1.default;
const Type_1 = __importDefault(require("./Type"));
exports.Type = Type_1.default;
const Ui_1 = __importDefault(require("./Ui"));
exports.Ui = Ui_1.default;
const UiTransaction_1 = __importDefault(require("./UiTransaction"));
exports.UiTransaction = UiTransaction_1.default;
const User_1 = __importDefault(require("./User"));
exports.User = User_1.default;
const WordKey_1 = __importDefault(require("./WordKey"));
exports.WordKey = WordKey_1.default;
const Currency_1 = __importDefault(require("./Currency"));
exports.Currency = Currency_1.default;
const Location_1 = __importDefault(require("./Location"));
exports.Location = Location_1.default;
exports.entities = {
    AccountStatus: AccountStatus_1.default,
    AccountStatusType: AccountStatusType_1.default,
    Branch: Branch_1.default,
    Catalog: Catalog_1.default,
    CatalogValue: CatalogValue_1.default,
    Chat: Chat_1.default,
    ChatType: ChatType_1.default,
    ChatUser: ChatUser_1.default,
    Config: Config_1.default,
    ConfigValue: ConfigValue_1.default,
    File: File_1.default,
    FileType: FileType_1.default,
    GlobalData: GlobalData_1.default,
    Group: Group_1.default,
    HistoryFile: HistoryFile_1.default,
    Language: Language_1.default,
    Log: Log_1.default,
    Message: Message_1.default,
    Notification: Notification_1.default,
    Person: Person_1.default,
    Report: Report_1.default,
    ReportGroup: ReportGroup_1.default,
    Role: Role_1.default,
    ScriptVersion: ScriptVersion_1.default,
    Session: Session_1.default,
    Subsystem: Subsystem_1.default,
    Subtype: Subtype_1.default,
    Template: Template_1.default,
    Transaction: Transaction_1.default,
    Translate: Translate_1.default,
    TranslationGroup: TranslationGroup_1.default,
    Type: Type_1.default,
    Ui: Ui_1.default,
    UiTransaction: UiTransaction_1.default,
    User: User_1.default,
    WordKey: WordKey_1.default,
    Currency: Currency_1.default,
    Location: Location_1.default
};
