// @pxp-nd/common/dist/controllers/Catalog.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const typeorm_1 = require("typeorm");
const core_1 = require("@pxp-nd/core");
const core_2 = require("@pxp-nd/core");
const entities_1 = require("../entities");
let Catalog = class Catalog extends core_1.Controller {
    async find(params) {
        const [data, count] = await typeorm_1.getManager().findAndCount(entities_1.Catalog, this.pxpParams);
        return { data, count };
    }
    async findById(params) {
        const catalog = await typeorm_1.getManager().findOne(entities_1.Catalog, {
            where: { catalogId: this.pxpParams.id }
        });
        if (!catalog) {
            return {};
        }
        else {
            return {
                catalog,
            };
        }
    }
    async save(params) {
        const catalog = await typeorm_1.getManager().save(entities_1.Catalog, Object.assign({}, this.pxpParams));
        return catalog;
    }
    async edit(params) {
        console.log(params);
        const id = this.pxpParams.id;
        delete this.pxpParams.id;
        await typeorm_1.getManager().update(entities_1.Catalog, id, Object.assign({}, this.pxpParams));
        const catalog = await typeorm_1.getManager().findOne(entities_1.Catalog, {
            where: { catalogId: id }
        });
        return catalog;
    }
    async delete(params) {
        console.log('params', this.pxpParams);
        const catalog = await typeorm_1.getManager().delete(entities_1.Catalog, {
            catalogId: this.pxpParams.id
        });
        return catalog;
    }
};
__decorate([
    core_2.Get(),
    core_2.DbSettings('Orm'),
    core_2.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Catalog.prototype, "find", null);
__decorate([
    core_2.Get('/find/:id'),
    core_2.DbSettings('Orm'),
    core_2.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Catalog.prototype, "findById", null);
__decorate([
    core_2.Post(),
    core_2.DbSettings('Orm'),
    core_2.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Catalog.prototype, "save", null);
__decorate([
    core_2.Patch('/edit/:id'),
    core_2.DbSettings('Orm'),
    core_2.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Catalog.prototype, "edit", null);
__decorate([
    core_2.Delete('/delete/:id'),
    core_2.DbSettings('Orm'),
    core_2.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Catalog.prototype, "delete", null);
Catalog = __decorate([
    core_2.Route('/catalogs')
], Catalog);
exports.default = Catalog;
