// @pxp-nd/common/dist/controllers/index1.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WordKey = exports.User = exports.Ui = exports.Type = exports.TranslationGroup = exports.Translate = exports.Template = exports.Person = exports.Language = exports.CatalogValue = exports.Catalog = exports.AccountStatusType = exports.AccountStatus = void 0;
const AccountStatus_1 = __importDefault(require("./AccountStatus"));
exports.AccountStatus = AccountStatus_1.default;
var AccountStatusType_1 = require("./AccountStatusType");
Object.defineProperty(exports, "AccountStatusType", { enumerable: true, get: function () { return __importDefault(AccountStatusType_1).default; } });
var Catalog_1 = require("./Catalog");
Object.defineProperty(exports, "Catalog", { enumerable: true, get: function () { return __importDefault(Catalog_1).default; } });
var CatalogValue_1 = require("./CatalogValue");
Object.defineProperty(exports, "CatalogValue", { enumerable: true, get: function () { return __importDefault(CatalogValue_1).default; } });
var Language_1 = require("./Language");
Object.defineProperty(exports, "Language", { enumerable: true, get: function () { return __importDefault(Language_1).default; } });
var Person_1 = require("./Person");
Object.defineProperty(exports, "Person", { enumerable: true, get: function () { return __importDefault(Person_1).default; } });
var Template_1 = require("./Template");
Object.defineProperty(exports, "Template", { enumerable: true, get: function () { return __importDefault(Template_1).default; } });
var Translate_1 = require("./Translate");
Object.defineProperty(exports, "Translate", { enumerable: true, get: function () { return __importDefault(Translate_1).default; } });
var TranslationGroup_1 = require("./TranslationGroup");
Object.defineProperty(exports, "TranslationGroup", { enumerable: true, get: function () { return __importDefault(TranslationGroup_1).default; } });
var Type_1 = require("./Type");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return __importDefault(Type_1).default; } });
var Ui_1 = require("./Ui");
Object.defineProperty(exports, "Ui", { enumerable: true, get: function () { return __importDefault(Ui_1).default; } });
var User_1 = require("./User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return __importDefault(User_1).default; } });
var WordKey_1 = require("./WordKey");
Object.defineProperty(exports, "WordKey", { enumerable: true, get: function () { return __importDefault(WordKey_1).default; } });
