// @pxp-nd/common/dist/controllers/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.controllers = exports.WordKeyController = exports.UserController = exports.UiController = exports.TypeController = exports.TranslationGroupController = exports.TranslateController = exports.TemplateController = exports.PersonController = exports.LanguageController = exports.ConfigController = exports.CatalogValueController = exports.CatalogController = exports.AccountStatusTypeController = exports.AccountStatusController = void 0;
const AccountStatus_1 = __importDefault(require("./AccountStatus"));
exports.AccountStatusController = AccountStatus_1.default;
const AccountStatusType_1 = __importDefault(require("./AccountStatusType"));
exports.AccountStatusTypeController = AccountStatusType_1.default;
const Catalog_1 = __importDefault(require("./Catalog"));
exports.CatalogController = Catalog_1.default;
const CatalogValue_1 = __importDefault(require("./CatalogValue"));
exports.CatalogValueController = CatalogValue_1.default;
const Config_1 = __importDefault(require("./Config"));
exports.ConfigController = Config_1.default;
const Language_1 = __importDefault(require("./Language"));
exports.LanguageController = Language_1.default;
const Person_1 = __importDefault(require("./Person"));
exports.PersonController = Person_1.default;
const Template_1 = __importDefault(require("./Template"));
exports.TemplateController = Template_1.default;
const Translate_1 = __importDefault(require("./Translate"));
exports.TranslateController = Translate_1.default;
const TranslationGroup_1 = __importDefault(require("./TranslationGroup"));
exports.TranslationGroupController = TranslationGroup_1.default;
const Type_1 = __importDefault(require("./Type"));
exports.TypeController = Type_1.default;
const Ui_1 = __importDefault(require("./Ui"));
exports.UiController = Ui_1.default;
const User_1 = __importDefault(require("./User"));
exports.UserController = User_1.default;
const WordKey_1 = __importDefault(require("./WordKey"));
exports.WordKeyController = WordKey_1.default;
exports.controllers = [
    AccountStatus_1.default,
    AccountStatusType_1.default,
    Catalog_1.default,
    CatalogValue_1.default,
    Config_1.default,
    Language_1.default,
    Person_1.default,
    Template_1.default,
    Translate_1.default,
    TranslationGroup_1.default,
    Type_1.default,
    Ui_1.default,
    User_1.default,
    WordKey_1.default,
];
