// @pxp-nd/common/dist/controllers/WordKey.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Kplian Ltda 2020
 *
 * MIT
 *
 * User Controller
 *
 * @summary User Controller
 * @author Israel Cm
 *
 * Created at     : 2020-09-17 18:55:38
 * Last modified  : 2020-09-17 19:04:30
 */
const typeorm_1 = require("typeorm");
const core_1 = require("@pxp-nd/core");
const entities_1 = require("../entities");
let WordKey = class WordKey extends core_1.Controller {
    async add(params, manager) {
        const wordKey = await manager.save(entities_1.WordKey, Object.assign({}, params));
        return wordKey;
    }
    async findById(params, manager) {
        const [data, count] = await manager.findAndCount(entities_1.WordKey, {
            where: {
                translationGroupId: params.groupId,
            },
            relations: ['group']
        });
        return { data, count };
    }
    async list(params) {
        const [data, count] = await typeorm_1.getManager().findAndCount(entities_1.WordKey, Object.assign(Object.assign({}, params), { relations: ['group'] }));
        const parseData = data.map(item => (Object.assign(Object.assign({}, item), { groupName: item.group ? item.group.name : '' })));
        return { data: parseData, count };
    }
};
__decorate([
    core_1.Get('/add'),
    core_1.DbSettings('Orm'),
    core_1.ReadOnly(false),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, typeorm_1.EntityManager]),
    __metadata("design:returntype", Promise)
], WordKey.prototype, "add", null);
__decorate([
    core_1.Get('/group/:groupId'),
    core_1.DbSettings('Orm'),
    core_1.ReadOnly(false),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, typeorm_1.EntityManager]),
    __metadata("design:returntype", Promise)
], WordKey.prototype, "findById", null);
__decorate([
    core_1.Get(),
    core_1.DbSettings('Orm'),
    core_1.ReadOnly(true),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], WordKey.prototype, "list", null);
WordKey = __decorate([
    core_1.Route('/translate/words'),
    core_1.Model('common/WordKey')
], WordKey);
exports.default = WordKey;
