// @pxp-nd/auth/dist/config/passport-facebook.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.configFacebookStrategy = void 0;
/**
 * Kplian Ltda 2020
 *
 * MIT
 *
 * Facebook config
 *
 * @summary Facebook config
 * @author Israel Colque
 *
 * Created at     : 2020-06-13 18:09:48
 * Last modified  : 2020-09-17 18:44:52
 */
const passport_1 = __importDefault(require("passport"));
const passport_facebook_1 = require("passport-facebook");
const repositories_1 = require("@pxp-nd/repositories");
const typeorm_1 = require("typeorm");
const facebookStrategy = new passport_facebook_1.Strategy({
    clientID: String(process.env.FACE_CLIENT_ID),
    clientSecret: String(process.env.FACE_SECRET_ID),
    callbackURL: 'http://localhost:3000/auth/facebook/callback'
}, (accessToken, refreshToken, profile, done) => {
    const UserRepo = typeorm_1.getCustomRepository(repositories_1.UserRepository);
    UserRepo.findOrCreateSocial({
        authenticationId: profile.id,
        authenticationType: 'google'
    }, {
        username: profile.emails ? profile.emails[0].value : String(profile.id),
        authenticationId: String(profile.id),
        authenticationType: 'facebook'
    })
        .then((user) => {
        return done(null, user);
    })
        .catch((err) => done(err));
});
const configFacebookStrategy = () => {
    passport_1.default.use(facebookStrategy);
};
exports.configFacebookStrategy = configFacebookStrategy;
