// @pxp-nd/entities/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PxpEntity = exports.WordKey = exports.User = exports.UiTransaction = exports.Ui = exports.Type = exports.TranslationGroup = exports.Translate = exports.Template = exports.Subtype = exports.Subsystem = exports.Session = exports.ScriptVersion = exports.Role = exports.ReportGroup = exports.Report = exports.Person = exports.Notification = exports.Message = exports.Log = exports.Language = exports.HistoryFile = exports.GlobalData = exports.File = exports.ChatType = exports.Chat = exports.CatalogValue = exports.Catalog = exports.AccountStatusType = exports.AccountStatus = void 0;
var AccountStatus_1 = require("./entity/AccountStatus");
Object.defineProperty(exports, "AccountStatus", { enumerable: true, get: function () { return __importDefault(AccountStatus_1).default; } });
var AccountStatusType_1 = require("./entity/AccountStatusType");
Object.defineProperty(exports, "AccountStatusType", { enumerable: true, get: function () { return __importDefault(AccountStatusType_1).default; } });
var Catalog_1 = require("./entity/Catalog");
Object.defineProperty(exports, "Catalog", { enumerable: true, get: function () { return __importDefault(Catalog_1).default; } });
var CatalogValue_1 = require("./entity/CatalogValue");
Object.defineProperty(exports, "CatalogValue", { enumerable: true, get: function () { return __importDefault(CatalogValue_1).default; } });
var Chat_1 = require("./entity/Chat");
Object.defineProperty(exports, "Chat", { enumerable: true, get: function () { return __importDefault(Chat_1).default; } });
var ChatType_1 = require("./entity/ChatType");
Object.defineProperty(exports, "ChatType", { enumerable: true, get: function () { return __importDefault(ChatType_1).default; } });
var File_1 = require("./entity/File");
Object.defineProperty(exports, "File", { enumerable: true, get: function () { return __importDefault(File_1).default; } });
var GlobalData_1 = require("./entity/GlobalData");
Object.defineProperty(exports, "GlobalData", { enumerable: true, get: function () { return __importDefault(GlobalData_1).default; } });
var HistoryFile_1 = require("./entity/HistoryFile");
Object.defineProperty(exports, "HistoryFile", { enumerable: true, get: function () { return __importDefault(HistoryFile_1).default; } });
var Language_1 = require("./entity/Language");
Object.defineProperty(exports, "Language", { enumerable: true, get: function () { return __importDefault(Language_1).default; } });
var Log_1 = require("./entity/Log");
Object.defineProperty(exports, "Log", { enumerable: true, get: function () { return __importDefault(Log_1).default; } });
var Message_1 = require("./entity/Message");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return __importDefault(Message_1).default; } });
var Notification_1 = require("./entity/Notification");
Object.defineProperty(exports, "Notification", { enumerable: true, get: function () { return __importDefault(Notification_1).default; } });
var Person_1 = require("./entity/Person");
Object.defineProperty(exports, "Person", { enumerable: true, get: function () { return __importDefault(Person_1).default; } });
var Report_1 = require("./entity/Report");
Object.defineProperty(exports, "Report", { enumerable: true, get: function () { return __importDefault(Report_1).default; } });
var ReportGroup_1 = require("./entity/ReportGroup");
Object.defineProperty(exports, "ReportGroup", { enumerable: true, get: function () { return __importDefault(ReportGroup_1).default; } });
var Role_1 = require("./entity/Role");
Object.defineProperty(exports, "Role", { enumerable: true, get: function () { return __importDefault(Role_1).default; } });
var ScriptVersion_1 = require("./entity/ScriptVersion");
Object.defineProperty(exports, "ScriptVersion", { enumerable: true, get: function () { return __importDefault(ScriptVersion_1).default; } });
var Session_1 = require("./entity/Session");
Object.defineProperty(exports, "Session", { enumerable: true, get: function () { return __importDefault(Session_1).default; } });
var Subsystem_1 = require("./entity/Subsystem");
Object.defineProperty(exports, "Subsystem", { enumerable: true, get: function () { return __importDefault(Subsystem_1).default; } });
var Subtype_1 = require("./entity/Subtype");
Object.defineProperty(exports, "Subtype", { enumerable: true, get: function () { return __importDefault(Subtype_1).default; } });
var Template_1 = require("./entity/Template");
Object.defineProperty(exports, "Template", { enumerable: true, get: function () { return __importDefault(Template_1).default; } });
var Translate_1 = require("./entity/Translate");
Object.defineProperty(exports, "Translate", { enumerable: true, get: function () { return __importDefault(Translate_1).default; } });
var TranslationGroup_1 = require("./entity/TranslationGroup");
Object.defineProperty(exports, "TranslationGroup", { enumerable: true, get: function () { return __importDefault(TranslationGroup_1).default; } });
var Type_1 = require("./entity/Type");
Object.defineProperty(exports, "Type", { enumerable: true, get: function () { return __importDefault(Type_1).default; } });
var Ui_1 = require("./entity/Ui");
Object.defineProperty(exports, "Ui", { enumerable: true, get: function () { return __importDefault(Ui_1).default; } });
var UiTransaction_1 = require("./entity/UiTransaction");
Object.defineProperty(exports, "UiTransaction", { enumerable: true, get: function () { return __importDefault(UiTransaction_1).default; } });
var User_1 = require("./entity/User");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return __importDefault(User_1).default; } });
var WordKey_1 = require("./entity/WordKey");
Object.defineProperty(exports, "WordKey", { enumerable: true, get: function () { return __importDefault(WordKey_1).default; } });
var PxpEntity_1 = require("./PxpEntity");
Object.defineProperty(exports, "PxpEntity", { enumerable: true, get: function () { return PxpEntity_1.PxpEntity; } });
